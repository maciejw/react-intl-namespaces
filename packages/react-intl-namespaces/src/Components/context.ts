import * as PropTypes from 'prop-types';
import { IntlBackendProvider } from './IntlBackendProvider';

export interface TranslatedMessages {
  [key: string]: string;
}

export namespace IntlBackendContext {
  export type GetMessagesFromNamespace = (
    namespace: string,
    includeNamespace: string[],
  ) => void;

  export type AddMissingMessage = (message: MessageMetadata) => void;

  export type RegisterNamespaceDownloadNotification = (
    callback: (
      resource: { namespace: string; messages: TranslatedMessages },
    ) => void,
  ) => void;

  export type GetIntlProps = () => ReactIntl.IntlProvider.Props;
  export interface MessageMetadata {
    key: string;
    namespace: string;
    defaultMessage?: string;
    description?: string;
  }

  export interface DevConfigProps {
    includeMetadata: boolean;
    showIds: boolean;
  }
  interface IntlBackend extends DevConfigProps {
    getMessagesFromNamespace: GetMessagesFromNamespace;
    getIntlProps: GetIntlProps;
    addMissingMessage: AddMissingMessage;
    registerNamespaceDownloadNotification: RegisterNamespaceDownloadNotification;
  }
  export interface Context {
    intlBackend: IntlBackend;
  }

  const intlBackendShape: PropTypes.ValidationMap<IntlBackend> = {
    getMessagesFromNamespace: PropTypes.func.isRequired,
  };

  const intlContextTypes: React.ValidationMap<Context> = {
    intlBackend: PropTypes.shape(intlBackendShape),
  };

  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}

export namespace IntlNamespaceContext {
  interface IntlNamespace extends IntlBackendContext.DevConfigProps {
    getNameCurrentNamespace(): string;
    missingMessage(message: ReactIntl.FormattedMessage.MessageDescriptor): void;
  }
  export interface Context {
    intlNamespace: IntlNamespace;
  }
  const intlNamespaceShape: PropTypes.ValidationMap<IntlNamespace> = {
    missingMessage: PropTypes.func.isRequired,
  };
  const intlContextTypes: React.ValidationMap<Context> = {
    intlNamespace: PropTypes.shape(intlNamespaceShape),
  };

  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}
