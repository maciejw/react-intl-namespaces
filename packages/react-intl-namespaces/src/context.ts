import * as PropTypes from 'prop-types';
import { ResourceFromNamespace } from './types';

export namespace IntlBackendContext {
  export type GetMessagesFromNamespace = (
    namespaceLoadedNotification: (resource: ResourceFromNamespace) => void,
    namespace: string,
    includeNamespace: string[],
  ) => void;

  export type AddMissingMessage = (message: MessageMetadata) => void;

  export type GetIntlProps = () => ReactIntl.IntlProvider.Props;
  export interface MessageMetadata {
    key: string;
    namespace: string;
    defaultMessage?: string;
    description?: string;
  }

  export interface DevConfigProps {
    includeMetadata: boolean;
    showIds: boolean;
    loggingEnabled: boolean;
  }
  export interface IntlBackend extends DevConfigProps {
    getMessagesFromNamespace: GetMessagesFromNamespace;
    getIntlProps: GetIntlProps;
    addMissingMessage: AddMissingMessage;
  }
  export interface Context {
    intlBackend: IntlBackend;
  }

  const intlBackendShape: PropTypes.ValidationMap<IntlBackend> = {
    getMessagesFromNamespace: PropTypes.func.isRequired,
  };

  const intlContextTypes: React.ValidationMap<Context> = {
    intlBackend: PropTypes.shape(intlBackendShape),
  };
  /**
   * decorator function that defines consumer end of react context
   * @param target component that acquires context
   */
  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  /**
   * decorator function that defines publisher end of react context
   * @param target component that creates context
   */
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}

export namespace IntlNamespaceContext {
  export interface IntlNamespace extends IntlBackendContext.DevConfigProps {
    getNameOfCurrentNamespace(): string;
    missingMessage(message: ReactIntl.FormattedMessage.MessageDescriptor): void;
  }
  export interface Context {
    intlNamespace: IntlNamespace;
  }
  const intlNamespaceShape: PropTypes.ValidationMap<IntlNamespace> = {
    missingMessage: PropTypes.func.isRequired,
  };
  export const intlContextTypes: React.ValidationMap<Context> = {
    intlNamespace: PropTypes.shape(intlNamespaceShape),
  };

  /**
   * decorator function that defines consumer end of react context
   * @param target component that acquires context
   */
  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  /**
   * decorator function that defines publisher end of react context
   * @param target component that creates context
   */
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}
