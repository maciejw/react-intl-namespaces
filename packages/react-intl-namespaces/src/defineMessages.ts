import { Messages } from './types';

/**
 * defines ReactIntl messages object that works with intellisense and type checking in typescript. It defines on the fly type, based on names of properties passed as a generic parameter
 * @param messages type safe object with messages
 * @returns passed messages object
 */
export function defineMessages<T extends string>(
  messages: Messages<T>,
): Messages<T>;
export function defineMessages(messages: any) {
  return messages;
}
