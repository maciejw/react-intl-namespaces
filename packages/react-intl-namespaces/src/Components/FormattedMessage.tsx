import invariant from 'invariant';
import * as React from 'react';
import * as ReactIntl from 'react-intl';
import { IntlNamespaceContext } from '../context';
import { IntlNamespaces } from '../namespaces';

/** */
@IntlNamespaceContext.Provide
export class FormattedMessage extends ReactIntl.FormattedMessage {
  // prettier-ignore
  public context!: IntlNamespaceContext.Context;
  constructor(
    props: ReactIntl.FormattedMessage.Props,
    context: IntlNamespaceContext.Context,
  ) {
    invariant(
      context.intlNamespace !== undefined,
      'Missing intlNamespace context. Use IntlNamespaceProvider inside IntlBackendProvider',
    );
    super(props, context);
  }

  public render() {
    const result = super.render();

    if (this.context.intlNamespace.includeMetadata) {
      const currentNamespace = this.context.intlNamespace.getNameOfCurrentNamespace();
      const metadata = IntlNamespaces.getMessageMetadata(
        this.props,
        currentNamespace,
      );

      return React.createElement(
        'span',
        {
          'data-default-message': metadata.defaultMessage,
          'data-description': metadata.description,
          'data-key': metadata.key,
          'data-namespace': metadata.namespace,
        },
        result,
      );
    } else {
      return result;
    }
  }
}
