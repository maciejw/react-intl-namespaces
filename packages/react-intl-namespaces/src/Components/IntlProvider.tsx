import invariant from 'invariant';
import * as React from 'react';
import * as ReactIntl from 'react-intl';
import { IntlNamespaceContext } from '../context';
import { IntlNamespaces } from '../namespaces';
import { NamespaceResource } from '../types';

@IntlNamespaceContext.Provide
export class IntlProvider extends ReactIntl.IntlProvider {
  // prettier-ignore
  public context!: IntlNamespaceContext.Context;
  constructor(props: ReactIntl.IntlProvider.Props, context: {}) {
    super(props, context);
  }

  public getChildContext() {
    const result = super.getChildContext();

    const { formatMessage: intlFormatMessage, ...rest } = result.intl;

    invariant(
      this.context.intlNamespace !== undefined,
      'Missing intlNamespace context. Use IntlNamespaceProvider inside IntlBackendProvider',
    );

    const {
      getNameOfCurrentNamespace,
      missingMessage,
      showIds,
    } = this.context.intlNamespace;

    const formatMessage = (
      messageDescriptor: ReactIntl.FormattedMessage.MessageDescriptor,
      values: { [key: string]: ReactIntl.MessageValue } = {},
    ) => {
      if (!this.hasResource(messageDescriptor.id)) {
        missingMessage(messageDescriptor);
      }

      if (showIds) {
        return IntlNamespaces.formatResourceId(
          messageDescriptor,
          getNameOfCurrentNamespace(),
          Object.getOwnPropertyNames(values),
        );
      } else {
        return intlFormatMessage(messageDescriptor, values);
      }
    };

    return {
      intl: {
        formatMessage,
        ...rest,
      },
    };
  }
  public render() {
    return super.render();
  }
  private hasResource(id: string) {
    const resource: NamespaceResource = this.props.messages || {};
    return resource.hasOwnProperty(id);
  }
}
