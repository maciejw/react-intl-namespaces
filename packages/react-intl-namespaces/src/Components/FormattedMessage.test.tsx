import * as PropTypes from 'prop-types';
import * as React from 'react';
import * as ReactIntl from 'react-intl';

import { mount } from 'enzyme';

import { IntlNamespaceContext } from '../context';
import { defineMessages } from '../defineMessages';
import { FormattedMessage } from './FormattedMessage';
import { IntlProvider } from './IntlProvider';

const messages = defineMessages<'property1' | 'property2'>({
  property1: {
    defaultMessage: 'text1 {someValue}',
    description: 'description1',
    id: 'property1',
  },
  property2: {
    defaultMessage: 'text2',
    description: 'description2',
    id: 'property2',
  },
});

const values = { someValue: 'value1' };
describe('FormattedMessage', () => {
  const sut = (
    <div>
      <FormattedMessage {...messages.property1} values={values} />
      <FormattedMessage {...messages.property2} />
    </div>
  );

  it('should display message', () => {
    const wrapper = mount(sut, {
      childContextTypes: {
        intl: PropTypes.object,
        intlNamespace: PropTypes.object,
      },
      context: {
        intl: {
          formatDate: jest.fn(),
          formatHTMLMessage: jest.fn(),
          formatNumber: jest.fn(),
          formatPlural: jest.fn(),
          formatRelative: jest.fn(),
          formatTime: jest.fn(),
          formatMessage(message: ReactIntl.FormattedMessage.MessageDescriptor) {
            return message.id;
          },
          now: jest.fn(),
          textComponent: 'span',
        },
        intlNamespace: {
          includeMetadata: true,
          getNameOfCurrentNamespace() {
            return 'namespace';
          },
          missingMessage: jest.fn(),
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
