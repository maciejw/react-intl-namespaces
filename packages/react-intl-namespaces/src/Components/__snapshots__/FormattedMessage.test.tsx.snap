// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedMessage should display message 1`] = `"<div><span data-default-message=\\"text1 {someValue}\\" data-description=\\"description1\\" data-key=\\"property1\\" data-namespace=\\"namespace\\"><span>property1</span></span><span data-default-message=\\"text2\\" data-description=\\"description2\\" data-key=\\"property2\\" data-namespace=\\"namespace\\"><span>property2</span></span></div>"`;
